vignettes/creating_linters.Rmd

```diff
@@ -11,113 +11,186 @@ vignette: >
-it is recommended to work with the tokens from `source_file$parsed_content` if
-possible, as they are tokenized from the `R` parser. These tokens are obtained
-from `parse()` and `getParseData()` calls done prior to calling the new linter.
+Therefore it is recommended to work with the tokens from `source_file$parsed_content` or `source_file$xml_parsed_content`,
+as they are tokenized from the `R` parser. These tokens are obtained
+from `parse()` and `utils::getParseData()` calls done prior to calling the new linter.
 `getParseData()` returns a `data.frame` with information from the source parse
 tree of the file being linted. A list of tokens is available from
-[r-source/src/main/gram.y](https://github.com/wch/r-source/blob/ff1bca2f21aba271d428474f00494eece5c36dd3/src/main/gram.y#L293-L307).
+[r-source/src/main/gram.y](https://github.com/r-devel/r-svn/blob/master/src/main/gram.y#L395-L412).
+
+`source_file$xml_parsed_content` uses `xmlparsedata::xml_parse_data()` to convert the `getParseData()` output into an XML tree,
+which enables writing linter logic in [XPath](https://www.w3schools.com/xml/xpath_intro.asp), a powerful language
+for expressing paths within the nested XML data structure. Most linters in `lintr` are built using XPath because it is
+concise to write and fast to execute.
 
 ```r
-lapply(ids_with_token(source_file, "EQ_ASSIGN"),
-  function(id) {
-    parsed <- with_id(source_file, id)
+if (!is_lint_level(source_expression, "expression", require_xml = TRUE)) {
+  return(list())
+}
 ```
-`ids_with_token()` is used to search for a token of a specific type and return
-the associated `id`.  Note that `rownames()` for `parsed_content` returns this
-`id`, so you can retrieve the rows for a given id with
-`source_file$parsed_content[id, ]`. The same operation may be more conveniently
-carried out with `with_id(source_file, id)`.
+
+Here, we return early if `source_expression` is not the expression-level object. `get_source_expression()` returns an
+object that parses the input file in two ways -- once is done expression-by-expression, the other contains all of the
+expressions in the file. This is done to facilitate caching. Suppose your package has a long source file (e.g., 100s of expressions) --
+rather than run linters on every expression every time the file is updated, when caching is activated `lintr` will
+only run the linter again on expressions that have changed.
+
+Therefore, it is preferable to write expression-level linters whenever possible. Two types of exceptions observed in `lintr` are
+(1) when several or all expressions are _required_ to ensure the linter logic applies (e.g., `conjunct_test_linter` looks for
+consecutive calls to `stopifnot()`, which will typically appear on consecutive expressions) or (2) when the linter logic
+is very simple & fast to compute, so that the overhead of re-running the linter is low (e.g., `single_quotes_linter`).
```

Comment:
Perhaps add a paragraph explaining the purpose of `require_xml = TRUE` and note that it's only required if XPath is used instead of parsed_content directly?

Reply:
actually, let's just get rid of the need for this altogether...

Reply:
What do you mean? It was a reaction to `xmlparsedata` bugs in the past, where `parsed_content` was still fine but the XML was invalid.
Or do you think we should no longer run parsed_content based linters if the XML is invalid?

Reply:
Let's discuss in #1147

Reply:
OK it looks like #1147 will land. I'll adjust the wording here quickly and then we can proceed.